### PR TITLE
rocm-related: 5.3.1 -> 5.3.3 & misc changes

### PR DESCRIPTION
**Pull request by Madouura**
###### Description of changes
Tracking: #197885

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Madouura** [comment]
Looks like I forgot ``composable_kernel`` and ``rocmlir`` - whoops.

**Comment by Madouura** [comment]
Result of `nixpkgs-review pr 202373` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>clang-ocl</li>
    <li>composable_kernel</li>
    <li>hip</li>
    <li>hipcub</li>
    <li>hipsparse</li>
    <li>llvmPackages_rocm.clang</li>
    <li>llvmPackages_rocm.clangNoCompilerRt</li>
    <li>llvmPackages_rocm.llvm</li>
    <li>miopengemm</li>
    <li>rccl</li>
    <li>rocblas</li>
    <li>rocclr</li>
    <li>rocfft</li>
    <li>rocm-cmake</li>
    <li>rocm-comgr</li>
    <li>rocm-device-libs</li>
    <li>rocm-opencl-icd</li>
    <li>rocm-opencl-runtime</li>
    <li>rocm-runtime</li>
    <li>rocm-smi</li>
    <li>rocm-thunk</li>
    <li>rocminfo</li>
    <li>rocmlir</li>
    <li>rocprim</li>
    <li>rocsparse</li>
    <li>rocthrust</li>
    <li>tensile</li>
  </ul>
</details>

**Review by Flakebi — APPROVED** [review]
Nice!
I tried some OpenCL and ROCm examples and they worked fine.

I found a way to call the update script with pressing enter only a single time: `nix-shell maintainers/scripts/update.nix --argstr commit true --argstr keep-going true --arg predicate '(path: pkg: builtins.elem (pkg.pname or null) [ "clang-ocl" "composable_kernel" "hip" "hipcub" "hipsparse" "miopengemm" "rccl" "rocblas" "rocclr" "rocfft" "rocprim" "rocsparse" "rocthrust" "rocm-cmake" "rocm-comgr" "rocm-device-libs" "rocm-llvm" "rocm-opencl-runtime" "rocm-runtime" "rocm-smi" "rocm-thunk" "rocminfo" "tensile" ])'`

**Comment by Madouura** [comment]
Niiiiice. :+1:

**Comment by Madouura** [comment]
In my TODO I have:
- Change all relevant derivations from ``rec`` to ``finalAttrs``
- Change all relevant derivation's maintainers to rocm team + previous non-team maintainers
Should I include them in this PR? And if I do, should I roll all of these changes into one commit per derivation?
On one hand the flexibility of just being able to revert each commit if needed is nice, on the other hand it feels like commit spam.

**Comment by acowley** [comment]
I think one commit per listed change to the package set, not one commit per derivation, is fine.

**Comment by Madouura** [comment]
Not using ``finalAttrs`` for ``tensile``  because it doesn't work when used as an optional ``null`` for another derivation input.

**Comment by Madouura** [comment]
Result of `nixpkgs-review pr 202373` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>27 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>clang-ocl</li>
    <li>composable_kernel</li>
    <li>hip</li>
    <li>hipcub</li>
    <li>hipsparse</li>
    <li>llvmPackages_rocm.clang</li>
    <li>llvmPackages_rocm.clangNoCompilerRt</li>
    <li>llvmPackages_rocm.llvm</li>
    <li>miopengemm</li>
    <li>rccl</li>
    <li>rocblas</li>
    <li>rocclr</li>
    <li>rocm-cmake</li>
    <li>rocm-comgr</li>
    <li>rocm-device-libs</li>
    <li>rocm-opencl-icd</li>
    <li>rocm-opencl-runtime</li>
    <li>rocm-runtime</li>
    <li>rocm-smi</li>
    <li>rocm-thunk</li>
    <li>rocminfo</li>
    <li>rocmlir</li>
    <li>rocprim</li>
    <li>rocsparse</li>
    <li>rocthrust</li>
    <li>tensile</li>
  </ul>
</details>

**Comment by Madouura** [comment]
Only change was ``rocmlir`` using the now-added tags. https://github.com/ROCmSoftwarePlatform/rocMLIR/tags

File: pkgs/development/compilers/llvm/rocm/default.nix [hunk 1]
```diff
@@ -1,7 +1,7 @@
 { stdenv, lib, buildPackages, fetchFromGitHub, callPackage, wrapCCWith, overrideCC }:
 
 let
-  version = "5.3.1";
+  version = "5.3.3";
```
Comment (mweinelt) [inline pkgs/development/compilers/llvm/rocm/default.nix:4]: version changed, but hash stayed the same?

File: pkgs/development/libraries/clang-ocl/default.nix [hunk 2]
```diff
@@ -1,21 +1,22 @@
...
-  version = rocmVersion;
+  rocmVersion = "5.3.3";
+  version = finalAttrs.rocmVersion;
 
   src = fetchFromGitHub {
     owner = "RadeonOpenCompute";
     repo = "clang-ocl";
-    rev = "rocm-${rocmVersion}";
+    rev = "rocm-${finalAttrs.rocmVersion}";
     hash = "sha256-uMSvcVJj+me2E+7FsXZ4l4hTcK6uKEegXpkHGcuist0=";
```
Comment (mweinelt) [inline pkgs/development/libraries/clang-ocl/default.nix:20]: version change, but no hash change?

File: pkgs/development/compilers/llvm/rocm/default.nix (hunk 1 above)
Comment (Flakebi) [reply pkgs/development/compilers/llvm/rocm/default.nix:4]: Yes, this happens quite often with ROCm. All projects get a new tag for a minor version, but only some of them have actual changes.

**Comment by mweinelt** [comment]
Result of `nixpkgs-review pr 202373` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>27 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>clang-ocl</li>
    <li>composable_kernel</li>
    <li>hip</li>
    <li>hipcub</li>
    <li>hipsparse</li>
    <li>llvmPackages_rocm.clang</li>
    <li>llvmPackages_rocm.clangNoCompilerRt</li>
    <li>llvmPackages_rocm.llvm</li>
    <li>miopengemm</li>
    <li>rccl</li>
    <li>rocblas</li>
    <li>rocclr</li>
    <li>rocm-cmake</li>
    <li>rocm-comgr</li>
    <li>rocm-device-libs</li>
    <li>rocm-opencl-icd</li>
    <li>rocm-opencl-runtime</li>
    <li>rocm-runtime</li>
    <li>rocm-smi</li>
    <li>rocm-thunk</li>
    <li>rocminfo</li>
    <li>rocmlir</li>
    <li>rocprim</li>
    <li>rocsparse</li>
    <li>rocthrust</li>
    <li>tensile</li>
  </ul>
</details>

File: pkgs/development/libraries/clang-ocl/default.nix [hunk 3]
```diff
@@ -34,11 +35,18 @@ stdenv.mkDerivation rec {
     "-DCMAKE_CXX_COMPILER=clang++"
   ];
 
+  passthru.updateScript = writeScript "update.sh" ''
+    #!/usr/bin/env nix-shell
+    #!nix-shell -i bash -p curl jq common-updater-scripts
+    rocmVersion="$(curl -sL "https://api.github.com/repos/RadeonOpenCompute/clang-ocl/releases?per_page=1" | jq '.[0].tag_name | split("-") | .[1]' --raw-output)"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/clang-ocl/default.nix:41]: This should optionally take a GITHUB_TOKEN to not run into rate limiting

Comment (Madouura) [reply pkgs/development/libraries/clang-ocl/default.nix:41]: Like this? https://github.com/NixOS/nixpkgs/blob/350fd0044447ae8712392c6b212a18bdf2433e71/pkgs/applications/emulators/rpcs3/update.sh

Comment (SuperSandro2000) [reply pkgs/development/libraries/clang-ocl/default.nix:41]: I was more thinking about something like this https://github.com/NixOS/nixpkgs/blob/b0f8c96b9ef5998f3985c34c90cfd21f1ba8a1ce/pkgs/applications/emulators/citra/update.sh#L30 to make it truly optional

Comment (Madouura) [reply pkgs/development/libraries/clang-ocl/default.nix:41]: Already done in the PR, so we should be good here.

**Pull request merged**